ambry-router/src/main/java/com.github.ambry.router/EncryptDecryptThread.java

```diff
@@ -0,0 +1,264 @@
+      this.callBack = callBack;
+    }
+
+    /**
+     * Steps to be performed on encryption
+     * 1. Encrypt blob content using perBlobKey
+     * 2. Fetch ContainerKey from kms for the given blob
+     * 3. Encrypt perBlobKey using containerKey
+     * 4. Invoke callback with the encryptedKey and encryptedContent
+     * @param cryptoService the {@link CryptoService} to assist in encrypting and decrypting
+     * @param kms the {@link KeyManagementService} to fetch keys from
+     */
+    @Override
+    void doOperation(CryptoService cryptoService, KeyManagementService kms) {
+      try {
+        ByteBuffer encryptedContent = cryptoService.encrypt(input, perBlobKey);
+        Object containerKey = kms.getKey(blobId.getAccountId(), blobId.getContainerId());
+        ByteBuffer encryptedKey = cryptoService.encryptKey(perBlobKey, containerKey);
+        callBack.onCompletion(blobId, encryptedKey, encryptedContent, null);
+      } catch (GeneralSecurityException e) {
+        callBack.onCompletion(blobId, null, null, e);
+      }
+    }
+
+    @Override
+    void closeJob(GeneralSecurityException gse) {
+      callBack.onCompletion(blobId, null, null, gse);
+    }
+  }
+
+  /**
+   * Class representing an decrypt Job.
+   */
+  static class DecryptJobInfo extends JobInfo {
+    private final ByteBuffer encryptedKey;
+    private final DecryptCallBack callBack;
+
+    /**
+     * Instantiates {@link DecryptJobInfo} with {@link BlobId}, key to be decrypted, content to be decrypted and the
+     * {@link DecryptCallBack}
+     * @param blobId the {@link BlobId} for which decryption is requested
+     * @param encryptedKey encrypted per blob key
+     * @param encryptedContent encrypted blob content
+     * @param callBack {@link DecryptCallBack} to be invoked on completion
+     */
+    DecryptJobInfo(BlobId blobId, ByteBuffer encryptedKey, ByteBuffer encryptedContent, DecryptCallBack callBack) {
+      super(blobId, encryptedContent);
+      this.encryptedKey = encryptedKey;
+      this.callBack = callBack;
+    }
+
+    /**
+     * Steps to be performed on decryption
+     * 1. Fetch ContainerKey from kms for the given blob
+     * 2. Decrypt encryptedKey using containerKey to obtain perBlobKey
+     * 3. Decrypt encryptedContent using perBlobKey
+     * 4. Invoke callback with the decryptedContent
+     * @param cryptoService the {@link CryptoService} to assist in encrypting and decrypting
+     * @param kms the {@link KeyManagementService} to fetch keys from
+     */
+    @Override
+    void doOperation(CryptoService cryptoService, KeyManagementService kms) {
+      try {
+        Object containerKey = kms.getKey(blobId.getAccountId(), blobId.getContainerId());
+        Object perBlobKey = cryptoService.decryptKey(encryptedKey, containerKey);
+        ByteBuffer decryptedContent = cryptoService.decrypt(input, perBlobKey);
+        callBack.onCompletion(blobId, decryptedContent, null);
+      } catch (GeneralSecurityException e) {
```

Comment:
If a runtime exception occurs here, we should probably notify the caller via the callback. On a related note, how do you feel about using the existing `Callback<T>` class and instead define `EncryptResult` and `DecryptResult` classes to use with it?

Reply:
Have changed the callback to catch Exception rather than GeneralSecurityException.